src/Airship/Internal/Helpers.hs

```diff
@@ -94,21 +98,73 @@ toWaiResponse Response{..} cfg trace quip =
-resourceToWai :: AirshipConfig -> RoutingSpec IO () -> Resource IO -> Wai.Application
-resourceToWai cfg routes resource404 =
-  resourceToWaiT cfg (const id) routes resource404
-
--- | Given a 'RoutingSpec', a 404 resource, and a user state @s@, construct a WAI 'Application'.
-resourceToWaiT :: Monad m => AirshipConfig -> (Request -> m Wai.Response -> IO Wai.Response) -> RoutingSpec m () -> Resource m -> Wai.Application
-resourceToWaiT cfg run routes resource404 req respond = do
+resourceToWai :: AirshipConfig
+              -> RoutingSpec IO ()
+              -> ErrorResponses IO
+              -> Wai.Application
+resourceToWai cfg routes errors =
+    resourceToWaiT cfg (const id) routes errors
+
+-- | Given a 'RoutingSpec', an 'ErrorResponses', and a user state @s@, construct a WAI 'Application'.
+resourceToWaiT :: Monad m =>
+                  AirshipConfig
+               -> (Request -> m Wai.Response -> IO Wai.Response)
+               -> RoutingSpec m ()
+               -> ErrorResponses m
+               -> Wai.Application
+resourceToWaiT cfg run routes errors req respond = do
     let routeMapping = runRouter routes
         pInfo = Wai.pathInfo req
-        (resource, (params', matched)) = route routeMapping pInfo resource404
-    nowTime <- getCurrentTime
     quip <- getQuip
-    (=<<) respond . run req $ do
-      (response, trace) <- eitherResponse nowTime params' matched req (flow resource)
-      return $ toWaiResponse response cfg (traceHeader trace) quip
+    nowTime <- getCurrentTime
+    case route routeMapping pInfo of
+        Nothing -> do
+            -- No route matched so bail out
+            let default404Resp = Response HTTP.status404 [(HTTP.hContentType, "text/plain")] Empty
+            respond =<< run req (do
+                (response, trace) <- eitherResponse nowTime mempty [] req (errorResponse errors default404Resp)
+                return $ toWaiResponse response cfg (traceHeader trace) quip)
+        Just (resource, (params', matched)) -> do
+            -- Merge the application-wide error responses with the resource
+            -- specific error responses favoring the resource-specific
+            --in case of conflict
+            let mergedResource = resource { errorResponses = M.union (errorResponses resource) errors }
+            respond =<< run req (do
+                (response, trace) <- eitherResponse nowTime params' matched req (flow mergedResource >>= errorResponse (errorResponses mergedResource))
+                return $ toWaiResponse response cfg (traceHeader trace) quip)
+
+
+-- | If the Response body is Empty the response body is set based on the error responses
+-- | provided by the application and resource. If the response body is not Empty or
+-- | there are no error response configured for the status code in the Response then no
+-- | action is taken. The contents of the 'Webmachine'' response body will be streamed
```

Comment:
Super nit, but the extra `|` on lines 2+ doesn't seem right. Not sure what haddock would do with them.